### PR TITLE
Irsa 4058 json i

### PR DESCRIPTION
**Pull request by judith-ipac**
The first commit just adds a pair of what used to be called #defines and does a global search-and-replace.
The second pulls out duplicated code from the ptree_reader::read_XXX() functions into Utils.[ch]xx files.  It also improves support for reading and writing IVOA-style links.
The third adds support for introducing arrays to hold duplicate keys, as Nelson suggested, when *writing* tables in JSON format, and also for extracting elements from such arrays while *reading* JSON-format tables and storing the elements in the usual place in the boost::property_tree structure used internally.

### CODE REVIEW

**Comment by judith-ipac** [comment]
Hi, Nelson.  I did a search and replace using "tablator::Labeled_Properties" as you suggested, and also added a STRING_PAIR typedef.  I made it all caps to be consistent with ATTRIBUTES, with which it frequently appears, as well as a raft of static consts. Will probably go back and make Labeled_Property/ies all caps as well.

**Comment by judith-ipac** [comment]
Hello again.  Remember when I tried to demonstrate a call to tablator and got this error message: 
`ptree contains data that cannot be represented in JSON format`?  Exactly as the message says, ptrees are allowed to store values in their nodes as well as (key, child_ptree) pairs, and that structure translates well to VOTABLE format, but it doesn't work for JSON format, in which nodes contain only (key, child) pairs and no stand-alone values.  Commit 8ef23597482 handles the problem by storing a node's value as the second component of a pair whose first component is the string "value" when preparing to write in JSON format.

File: src/Table/generate_property_tree/add_to_property_tree.cxx [hunk 1]
```diff
@@ -32,6 +32,29 @@ void Option_to_xml(boost::property_tree::ptree &tree, const tablator::Option &op
         for (auto &o : option.options) Option_to_xml(option_tree, o);
     }
 }
+
+
+// If json_prep is true, add a subtree with label <label> whether or
+// not a subtree already exists with that label.  Otherwise,
+// find---or, if none exists, create---a tree with label <label>_ARRAY
```
Comment (bonimy) [inline src/Table/generate_property_tree/add_to_property_tree.cxx:39]: `find (or, if none exists, create) a`

**Review by bonimy — APPROVED** [review]
👍

**Pull request merged**